### PR TITLE
Support tool and tool choice

### PR DESCRIPTION
**Pull request by 16oeahr**
# Description

Support tool and tool choice

- support role tool
- support assistant tool calls
- support validate tools and tool choice

Sample flow to use tools with chat model: https://msdata.visualstudio.com/Vienna/_git/PromptFlow/pullrequest/1315644

### CODE REVIEW

File: src/promptflow-tools/promptflow/tools/common.py [hunk 1]
```diff
@@ -82,35 +127,59 @@ def validate_functions(functions):
...
+              "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
+            },
+            "required": ["location"],
+          },
+        }
+      }
+    ]
+    """
+    tool_example = json.dumps(
+        {
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/common.py:157]: should have use `type` and `function` to wrap the example?

File: src/promptflow-tools/promptflow/tools/common.py [hunk 2]
```diff
@@ -123,10 +192,102 @@ def try_parse_name_and_content(role_prompt):
     return None
 
 
+def try_parse_tool_call_id_and_content(role_prompt):
+    # customer can add ## in front of name/content for markdown highlight.
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/common.py:196]: name -> tool_call_id?

File: src/promptflow-tools/promptflow/tools/common.py [hunk 3]
```diff
@@ -123,10 +192,102 @@ def try_parse_name_and_content(role_prompt):
...
+    # and we still support name/content without ## prefix for backward compatibility.
+    pattern = r"\n*#{0,2}\s*tool_call_id:\n+\s*(\S+)\s*\n*#{0,2}\s*content:\n?(.*)"
+    match = re.search(pattern, role_prompt, re.DOTALL)
+    if match:
+        return match.group(1), match.group(2)
+    return None
+
+
+def try_parse_tool_calls(role_prompt):
+    # customer can add ## in front of name/content for markdown highlight.
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/common.py:206]: name/content?

File: src/promptflow-tools/promptflow/tools/common.py [hunk 4]
```diff
@@ -123,10 +192,102 @@ def try_parse_name_and_content(role_prompt):
...
+    # and we still support name/content without ## prefix for backward compatibility.
+    pattern = r"## tool_calls:\n(.*)"
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/common.py:208]: why need to parse tool_calls? do not see this role in openai api ref.

File: src/promptflow-tools/promptflow/tools/common.py [hunk 5]
```diff
@@ -123,10 +192,102 @@ def try_parse_name_and_content(role_prompt):
...
+
+def is_tools_chunk(last_message):
+    return last_message and "role" in last_message and last_message["role"] == "tool" and "content" not in last_message
+
+
+def is_assistant_tool_calls_chunk(last_message, chunk):
+    return last_message and "role" in last_message and last_message["role"] == "assistant" and "tool_calls" in chunk
+
+
+def parse_tool_calls_for_assistant(last_message, chunk):
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/common.py:223]: can we postpone this support?

File: src/promptflow-tools/tests/test_common.py [hunk 6]
```diff
@@ -119,37 +258,250 @@ def test_success_parse_role_prompt(self, chat_str, images, expected_result):
...
+                        "tool_call_id": "call_001",
+                        "content": '{"location": "San Francisco, CA", "temperature": "72", "unit": null}',
+                    }
+                ],
+                True,
+                "",
+            ),
+            (
+                """
+# tool:
```
Comment (chjinche) [inline src/promptflow-tools/tests/test_common.py:291]: save in a separate file for this complicate test case.

File: src/promptflow-tools/tests/test_common.py [hunk 7]
```diff
@@ -119,37 +258,250 @@ def test_success_parse_role_prompt(self, chat_str, images, expected_result):
...
+                        "content": '{"location": "San Francisco, CA", "temperature": "72", "unit": null}',
+                    },
+                ],
+                True,
+                "",
+            ),
+            (
+                """
+# assistant:
+## tool_calls:
```
Comment (chjinche) [inline src/promptflow-tools/tests/test_common.py:332]: also this.

File: src/promptflow-tools/promptflow/tools/common.py (hunk 5 above)
Comment (16oeahr) [reply src/promptflow-tools/promptflow/tools/common.py:223]: offline synced

File: src/promptflow-tools/tests/test_common.py (hunk 6 above)
Comment (16oeahr) [reply src/promptflow-tools/tests/test_common.py]: updated

File: src/promptflow-tools/promptflow/tools/common.py (hunk 1 above)
Comment (16oeahr) [reply src/promptflow-tools/promptflow/tools/common.py:135]: fixed

File: src/promptflow-tools/tests/constants.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+                ],
+            }
+        ],
+        True,
+        "",
+        "",
+    ),
+    (
+        """
+            # system:
```
Comment (chjinche) [inline src/promptflow-tools/tests/constants.py:158]: Can we save it in jinja file and load in conftest? Then future tes cases can reuse the test prompt in the future?

image

Comment (16oeahr) [reply src/promptflow-tools/tests/constants.py]: moved

**Review by chjinche — APPROVED** [review]

**Pull request merged**